### ISSUE TITLE
Can't upgrade pnpm version:  EPERM: operation not permitted, unlink `AppData\Local\pnpm\pnpm.EXE`

### ISSUE THREAD

**Issue by Rolanddoda** [pnpm/pnpm#5700]
### Code to reproduce the issue:

I have only installed pnpm in my Windows 11 laptop because I want to manage node.js versions with it.
When I install dependencies in a project, I get an warning to upgrade pnpm to a newer version by running 

```bash
pnpm add -g @pnpm/exe
```

### Expected behavior:
Normally that command should work.

### Actual behavior:

After running that command I get:

![image](https://user-images.githubusercontent.com/18482346/204092510-e93b8f24-3099-4405-b7e9-307dda34a079.png)

The same happens with:

```bash
pnpm add -g pnpm
```

### Additional information:

 - `node -v` prints: v18.12.1
 - Windows, macOS, or Linux?: Windows 11 Pro

**Comment by florian-lefebvre** [comment]
Same here with Win11 and Node v16.15.0

**Comment by florian-lefebvre** [comment]
Yes and it didn't change anything. Let me know if you need me to provide more details about my env or to do some more testing.

**Comment by Rolanddoda** [comment]
> Have you tried to run your terminal with admin permissions?

Yes I did that, same error.

**Comment by zguiyang** [comment]
I have the same problem. Is there any temporary solution?

**Comment by zkochan** [comment]
Does the standalone script work to reinstall pnpm? 

https://pnpm.io/installation#on-windows

**Comment by Rolanddoda** [comment]
@zkochan this is what I got:

![image](https://user-images.githubusercontent.com/18482346/204774694-e2da06e3-3743-4cd1-90fc-13ad127c1106.png)

**Comment by Rolanddoda** [comment]
@zkochan yes that worked. 🎉

Does that mean that we have to open powershell and run:

```bash
iwr https://get.pnpm.io/install.ps1 -useb | iex
```

in order to upgrade pnpm?

**Comment by zkochan** [comment]
I think `pnpm add -g @pnpm/exe` should also be fixed. But use this upgrade method for now.

**Comment by Rolanddoda** [comment]
@zkochan please see my updated comment above! I updated it while you were adding the comment above

**Comment by ArcherGu** [comment]
Same problem 😢

**Comment by Deebster** [comment]
This issue also broke updating other packages:
```
PowerShell 7.3.1
PS C:\Users\micro> pnpm add -g wrangler
 WARN  deprecated rollup-plugin-inject@3.0.2: This package has been deprecated and is no longer maintained. Please use @rollup/plugin-inject.
 WARN  deprecated sourcemap-codec@1.4.8: Please use @jridgewell/sourcemap-codec instead
Packages: +462
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Progress: resolved 485, reused 461, downloaded 0, added 0, done
 WARN  The target bin directory already contains an exe called pnpm, so removing C:\Users\micro\AppData\Local\pnpm\pnpm.EXE
 EPERM  EPERM: operation not permitted, unlink 'C:\Users\micro\AppData\Local\pnpm\pnpm.EXE'


PS C:\Users\micro> pnpm -v
7.25.0
```

In the end I had to delete all C:\Users\micro\AppData\Local\pnpm* directories and reinstall from there.

**Comment by Deebster** [comment]
This issue still exists in 7.25.1:

```
PS C:\Users\micro> pnpm add -g @pnpm/exe

   ╭──────────────────────────────────────────────────────────────────╮
   │                                                                  │
   │                Update available! 7.25.1 → 7.26.3.                │
   │   Changelog: https://github.com/pnpm/pnpm/releases/tag/v7.26.3   │
   │         Run a script from: https://pnpm.io/installation          │
   │                                                                  │
   │      Follow @pnpmjs for updates: https://twitter.com/pnpmjs      │
   │                                                                  │
   ╰──────────────────────────────────────────────────────────────────╯

 WARN  deprecated rollup-plugin-inject@3.0.2: This package has been deprecated and is no longer maintained. Please use @rollup/plugin-inject.
 WARN  deprecated sourcemap-codec@1.4.8: Please use @jridgewell/sourcemap-codec instead
Packages: +2
++
Downloading registry.npmjs.org/@pnpm/win-x64/7.26.3: 17.4 MB/17.4 MB, done
Progress: resolved 496, reused 470, downloaded 2, added 2, done
.pnpm/@pnpm+exe@7.26.3/node_modules/@pnpm/exe: Running preinstall script, done in 299ms
 WARN  The target bin directory already contains an exe called pnpm, so removing C:\Users\micro\AppData\Local\pnpm\pnpm.EXE
 EPERM  EPERM: operation not permitted, unlink 'C:\Users\micro\AppData\Local\pnpm\pnpm.EXE'
```

**Comment by Sepruko** [comment]
I had this same issue, and it *also* prevented adding/upgrading **any** global package (not just `pnpm` itself, e.g. `turbo`).

I tried an alternative solution, which goes as follows...

1. Remove `%PNPM_HOME%\pnpm.exe` manually, e.g. through Windows' `explorer.exe` process.
2. Run `%PNPM_HOME%\global\5\.pnpm\@pnpm+exe@version\node_modules\@pnpm\exe\pnpm.exe add -g pnpm`.
3. ???
4. Profit.

> **Note**
> Powershell users will need to reference `PNPM_HOME` using `$Env:PNPM_HOME`.

After trying this, it works without a hitch. Don't come after me if it doesn't work for you though, I take no responsibility for any loss of life (caused by accidental nuclear meltdowns etc.) as a result of you attempting these steps.

![image](https://user-images.githubusercontent.com/33651534/223623827-facd4a29-d056-4692-9ebd-72a63ae58eb2.png)

**Comment by luke-hagar-sp** [comment]
This is a common issue on windows in my experience, with a number of different services, though for me its normally esbuild.

I would suggest hitting Ctrl + Shift + Esc and finding the running process, in this case its likely `pnpm.EXE` and then clicking `End task`

I hope this helps.

**Comment by Deebster** [comment]
@luke-hagar-sp but if you're using pnpm to update itself with the suggested command `pnpm add -g @pnpm/exe`, then pnpm.exe is _always_ running.

This used to work, whereas now it seems to always (or usually?) fail with an EPERM error.

**Comment by amihhs** [comment]
I ran into the same problem. After I encountered a strange problem I reinstalled pnpm using the following command

```
iwr https://get.pnpm.io/install.ps1 -useb | iex
```

Then I installed the latest pnpm@8.6.10

But when I use it, this problem occurs. When I execute `pnpm i -g`. I find that it reinstalls the pnpm@8.3.1 that I used before
I felt strange, so when I directly renamed pnpm.exe to another name, when I executed `pnpm -v` again it printed v8.3.1
After that it works fine

**Comment by adelarsq** [comment]
@amihhs That is a permission problem. Files on Windows are locked on use. So that explains why renaming pnpm.exe for another name works. But why its been blocked doesn't make sense.

**Comment by amihhs** [comment]
@adelarsq This problem is caused by the existence of an old version of pnpm and installing `sh` at the time of execution.
The execution of sh downloaded the new `pnpm.exe` file and moved it to the old working path. 
But when executing the pnpm setup command. Because the old pnpm still exists, the `pnpm setup` was not executed correctly (log: `No changes to the environment were made. Everything is already up to date.`), but the `pnpm.exe` file was not removed after the execution failed, and the same directory An older version of the pnpm file exists under .
When executing `pnpm [command]` next time, the pnpm.exe file will always be invoked instead of the correct pnpm executable. So that caused this error

**Comment by thakyZ** [comment]
If anyone is still having issue with this problem, might I suggest running `pnpm.cmd add -g pnpm @pnpm/exe` and see if that fixes things? (It did work for me, personally)

**Comment by Deebster** [comment]
@thakyZ I don't have pnpm.cmd, just npm.cmd, is that what you meant?

Either way, I've switched to [using winget](https://pnpm.io/installation#using-winget) to manage pnpm versions and so far it's worked flawlessly.

**Comment by thakyZ** [comment]
@Deebster I had the path to `PNPM_HOME` on my path so that I could use the `pnpm.cmd` from where ever `PNPM_HOME` is at. I don't think there should be a `npm.cmd` in that folder, but I assume the `PNPM_HOME` was not added to your path, and if that wasn't the cause I am not sure what I have different.

I installed pnpm thought npm and then just ran into the issue recently. I was just leaving it as a fix if it may be an alternative fix for anyone else, since I didn't see it covered or mentioned.

**Comment by ibqn** [comment]
I still have this error after installing `pnpm` with using powershell i.e. `iwr https://get.pnpm.io/install.ps1 -useb | iex`

```bash
$ pnpm --version
8.6.12

$ pnpm add -g pnpm
Nothing to stop. No server is running for the store at C:\Users\foobar\AppData\Local\pnpm\store\v3
Packages: +1
+
Progress: resolved 1, reused 1, downloaded 0, added 0, done
 WARN  The target bin directory already contains an exe called pnpm, so removing C:\Users\foobar\AppData\Local\pnpm\pnpm.EXE
 EPERM  EPERM: operation not permitted, unlink 'C:\Users\foobar\AppData\Local\pnpm\pnpm.EXE'
```

**Comment by andrekat** [comment]
> If anyone is still having issue with this problem, might I suggest running `pnpm.cmd add -g pnpm @pnpm/exe` and see if that fixes things? (It did work for me, personally)

solved the issue for me! thanks

**Comment by mo9a7i** [comment]
I believe due to admin rights restriction on my machine, I installed pnpm using `corepack enable` and I faced the issue after upgrading node to v18 and attempting to `pnpm -g update pnpm`. Not sure which one caused the issue, however, after installing node, I just `corepack enable` again then it worked.

**Comment by Elvincth** [comment]
fixed this problem after running `pnpm store prune` See: https://pnpm.io/cli/store

**Comment by kirbby** [comment]
> fixed this problem after running `pnpm store prune` See: https://pnpm.io/cli/store

This sadly changed nothing.
I used `pnpm add -g pnpm` all the time to update my pnpm until recently. Now it suddenly stopped working, this is a bit annoying, to always update via PowerShell (Windows 10 btw)

**Comment by itpropro** [comment]
Still no update?

**Comment by deeprobin** [comment]
I can repro

**Comment by huntermasonb** [comment]
After the update to 8.12 I ran into this issue. I was getting "operation not permitted, unlink 'C:\*\*\*\Local\pnpm\pnpm.EXE' ". I was able to work around this by renaming the 8.11 pnpm.exe file to "old-pnpm.exe", then `ran pnpm add -g pnpm` successfully.

I deleted the old .exe file and it seems to be fine for now, but I haven't restarted my PC yet so I'll find out once I do.

**Comment by MaxWilson** [comment]
I ran into this issue with an install via Powershell:

> iwr https://get.pnpm.io/install.ps1 -useb | iex

I deleted that install and installed via npm instead

> npm i -g pnpm

And the repro went away.

**Comment by Philboisvert** [comment]
Like @huntermasonb,

That's also the only thing that worked for me.
Renaming  `pnpm.exe` to something else then running `pnpm add -g pnpm`, it is indeed a permission issue.

**Comment by IIIMADDINIII** [comment]
TL;DR:
On Windows an executable which is executed can not be removed.
Depending on how pnpm was installed this breaks either the self update or the env command to change the Node version.
So a solution needs to be implemented possibly involving renaming the current running executable. 

I investigated this issue a little bit, because i wanted to use pnpm on Windows to manage my Node Version.
There are at least two ways pnpm can be installed on windows:
1. If you use the Installation Instructions on the Website (```iwr https://get.pnpm.io/install.ps1 -useb | iex```) you are left with a single binary in the users AppData directory (C:\Users\***\AppData\Local\pnpm\pnpm.exe).
2. If you install pnpm with the help of another Package Manager you are left with pnpm installed in a node_modules folder and the common pnpm shell files (pnpm, pnpm.cmd, pnpm.ps1) which will call the node.exe binary with pnpm in these shell folders.

If pnpm is installed as described in 1. you can not update pnpm with pnpm. If you execute ```pnpm add - pnpm``` or ```pnpm add - @pnpm/exe``` pnpm will try to delete the pnpm executable which is currently executed (pnpm is running right now).
With this installation of pnpm it is possible to manage the node version. You can run ```pnpm env use --global lts``` and pnpm will download and install node and npm in to the AppData directory. Also pnpm has no problem unlinking the node.exe file because pnpm is run with the pnpm.exe binary.

If pnpm is installed as described with 2. you can update pnpm by running ```pnpm add - pnpm```. Because now the update does not involve deleting the current running Binary (node.exe). pnpm only changes the files in the node_modules folder. In Windows point of view these files are normal text files which also have been read (but not locked) by node.exe. So they can be overridden.
If you try to change the node version with for example ```pnpm env use --global lts``` you will get the an error for the same reason pnpm could not update itself in the first case. Now again pnpm trys to delete (unlink) the file which is currently running (node.exe).

My proposal for a solution would be to create a special case for the node.exe and pnpm.exe files.
If the delete/unlink fails on these files, rename them to something like pnpm.exe.del instead.
If for some reason that file already exists use pnpm.exe.2.del or pnpm.exe.3.del and so on.
On subsequent runs of pnpm we could check for files with this naming scheme inside of the AppData directory and delete them.
This can be a special case, which is only used on windows.

I would like to hear feedback on this idea. Is there a better way to solve this chicken and egg problem?

**Comment by cwadrupldijjit** [comment]
Perhaps having a separate executable that is called by pnpm.exe as the final step that is fed the location of the new pnpm binary and watches for the current pnpm process to be terminated; at that point, the pnpm.exe binary is not locked by Windows running it and should be able to be deleted.

**Comment by cwadrupldijjit** [comment]
Having the separate binary be JS does not work for pnpm-managed node, however.  That runs afoul of node trying to overwrite itself, which Windows wouldn't like.  Consequently, a separate native-code binary might be best so that it doesn't conflict with the open node or pnpm processes.

**Comment by qiantum** [comment]
here is the solution i tested:
1. make sure node.js already installed
1. rename pnpm.exe to pnpm_.exe
2. pnpm_ -g i pnpm
3. make sure file pnpm.cmd appeared
4. delete pnpm_.exe
5. ok

**Comment by kirbby** [comment]
> [qiantum](/qiantum)

then it is easier to just use the command to install pnpm from the website (`iwr https://get.pnpm.io/install.ps1 -useb | iex`), instead of renaming and deleting.

**Comment by qiantum** [comment]
> then it is easier to just use the command to install pnpm from the website (`iwr https://get.pnpm.io/install.ps1 -useb | iex`), instead of renaming and deleting. I added an alias for that command in powershell and when i want to update ii open powershell with admin (on Windows) and execute the alias `updpnpm`. Which still annoys me, cause I need to open powershell as admin instead of just executing a command in VS Code. Probably the easiest would be, to just install pnpm with npm again to avoid those problems, like @IIIMADDINIII mentioned above.

i don't use any other managers, even npm. i completely excluded npm when installing nodejs

**Comment by ccthecode** [comment]
### Working on Windows

Each time I run `pnpm install -g @pnpm/exe` OR `pnpm add -g @pnpm/exe`. I keep getting this error

```
 WARN  The target bin directory already contains an exe called pnpm, so removing C:\Users\acces\AppData\Local\pnpm\pnpm.EXE
 EPERM  EPERM: operation not permitted, unlink 'C:\Users\acces\AppData\Local\pnpm\pnpm.EXE'
```

### SOULTION

✅ I finally got this to work on windows. (I initially renamed the original `pnpm.exe` manually to something different before running the command below but that may NOT be necessary. I believe just running the command below should work automatically)

```
curl -fsSL https://get.pnpm.io/install.sh | sh
```

It installed the latest version. Run the command below to check your latest version

```
pnpm -v
```

**Comment by goldingdamien** [comment]
This fixed this error message for me. Not sure if it is the same issue for others.
https://stackoverflow.com/a/61720199/1764521

**Comment by SetsuikiHyoryu** [comment]
I installed version 9.5.0 using winget on Windows 11 (it's my first time using pnpm), and encountered the same issue when upgrading to 9.6.0. Referring to some of the replies above, I renamed `pnpm.exe` to `pnpm1.exe`, then used `pnpm1` to execute `add` (see the code block below), and it was successfully installed. After that, I renamed it back to `pnpm.exe`.

```sh
hyoryu in ~
❯ pnpm add -g @pnpm/exe
Packages: +2
++
Progress: resolved 7, reused 2, downloaded 0, added 1, done
 WARN  The target bin directory already contains an exe called pnpm, so removing E:\SOFT\programming\package-manage\pnpm\pnpm.EXE
 EPERM  EPERM: operation not permitted, unlink 'E:\SOFT\programming\package-manage\pnpm\pnpm.EXE'


hyoryu in ~
❯ pnpm add -g @pnpm/exe
pnpm: The term 'pnpm' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
hyoryu in ~
❯ pnpm1 add -g @pnpm/exe
Packages: +2
++
Progress: resolved 7, reused 2, downloaded 0, added 0, done

E:\SOFT\programming\package-manage\pnpm\global_node_modules\5:
+ @pnpm/exe ^9.6.0

Done in 2.7s
hyoryu in ~
❯ pnpm --version
9.6.0
hyoryu in ~
❯

```

**Comment by imding** [comment]
> pnpm.cmd add -g pnpm @pnpm/exe

this worked for me

**Comment by sharky98** [comment]
Trying to update 9.9.0 to 9.11.0 today and got the same error using the provided code: `pnpm self-update`.

```console
~$ pnpm self-update
 WARN  The target bin directory already contains an exe called pnpm, so removing C:\Users\[me]\AppData\Local\pnpm\pnpm.EXE
 EPERM  EPERM: operation not permitted, unlink 'C:\Users\[me]\AppData\Local\pnpm\pnpm.EXE'
```

`pnpm store prune` changed nothing.

Doing the workaround by @qiantum above did work. However, that mean getting rid of the pnpm standalone executable. So far it doesn't seem to have any adverse effect.

**Comment by abcfy2** [comment]
I have the same issue on Windows. `pnpm self-update` failed. I have to use official windows install script to reinstall for upgrade:

```ps
PS C:\Users\abcfy> pnpm self-update
 WARN  The target bin directory already contains an exe called pnpm, so removing C:\Users\abcfy\AppData\Local\pnpm\pnpm.EXE
 EPERM  EPERM: operation not permitted, unlink 'C:\Users\abcfy\AppData\Local\pnpm\pnpm.EXE'


PS C:\Users\abcfy> pnpm --version
9.10.0
PS C:\Users\abcfy> Invoke-WebRequest https://get.pnpm.io/install.ps1 -UseBasicParsing | Invoke-Expression
Downloading pnpm from GitHub...

Running setup...

Copying pnpm CLI from C:\Users\abcfy\AppData\Local\Temp\84ad9c9d-6cec-4352-8b86-cf553a59c7d7\pnpm.exe to C:\Users\abcfy\AppData\Local\pnpm\pnpm.exe
No changes to the environment were made. Everything is already up to date.
PS C:\Users\abcfy> pnpm --version
9.11.0
PS C:\Users\abcfy>
```

**Comment by Rafael-Anguiano** [comment]
> @zkochan yes that worked. 🎉
> 
> Does that mean that we have to open powershell and run:
> 
> ```shell
> iwr https://get.pnpm.io/install.ps1 -useb | iex
> ```
> 
> in order to upgrade pnpm?
> 
> Because even on the latest version, the issue still persists: (after upgrading to the latest version)
> 
> ![image](https://user-images.githubusercontent.com/18482346/204775722-38248b71-fab1-4349-a156-4022626762e6.png)

This Worked!!!

**Comment by PMA365** [comment]
This worked for me. Just run it as Admin with cmd/Power shell both works. 😉

`npm install -g --force pnpm `

**Comment by deeprobin** [comment]
@gilbertoalbino AFAIK you could replace the file whilst running, because the executable is loaded into the memory. You just need to set the file share attributes.

**Comment by TexRx** [comment]
@PMA365 How does this work? My pnpm.exe is in `/AppData/Local/pnpm/pnpm`. Installing pnpm via `npm` installs `pnpm` in `C:\Users\H313383\AppData\Roaming\npm`. 

The only thing that worked for was running this in Powershell, based on @abcfy2 comment.
```
Invoke-WebRequest https://get.pnpm.io/install.ps1 -UseBasicParsing | Invoke-Expression
```

**Comment by nikhild791** [comment]
> [@PMA365](https://github.com/PMA365) How does this work? My pnpm.exe is in `/AppData/Local/pnpm/pnpm`. Installing pnpm via `npm` installs `pnpm` in `C:\Users\[username]\AppData\Roaming\npm`.
> 
> The only thing that worked for me was running the following in Powershell (admin), based on [@abcfy2](https://github.com/abcfy2) comment.
> 
> ```
> Invoke-WebRequest https://get.pnpm.io/install.ps1 -UseBasicParsing | Invoke-Expression
> ```

`C:\Users\admin\AppData\Local\pnpm` copy all the files from this dir to a new dir in this dir only oldPnpm
 then go to oldPnpm and run the cmd with admin prev check version of pnpm 
> ```
> pnpm -v
> ```
it will show the current version of it 

 now run 
> pnpm self-update 
there it will download the pnm latest version in the `C:\Users\admin\AppData\Local\pnpm` file 
> to check go the pnpm folder or any other than `C:\Users\admin\AppData\Local\pnpm\oldPnpm` it will show the latest pnpm

**Comment by you-hengh** [comment]
pnpm self-update  I've never seen such a foolish script command before. Was it even tested for success by the developers before being released on the Win11 Pro 24H2 environment?

![Image](https://github.com/user-attachments/assets/e7ce2abe-bbf5-449a-ae99-05f714ad3373)

![Image](https://github.com/user-attachments/assets/319ec209-8b56-4805-8058-1835f9d6b314)

**Comment by desarrollador1IR** [comment]
> pnpm self-update I've never seen such a foolish script command before. Was it even tested for success by the developers before being released on the Win11 Pro 24H2 environment?
> 
> ![Image](https://github.com/user-attachments/assets/e7ce2abe-bbf5-449a-ae99-05f714ad3373)
> 
> ![Image](https://github.com/user-attachments/assets/319ec209-8b56-4805-8058-1835f9d6b314)

i have the same error, is there a solution?:(

**Comment by you-hengh** [comment]
> > pnpm self-update I've never seen such a foolish script command before. Was it even tested for success by the developers before being released on the Win11 Pro 24H2 environment?
> > ![Image](https://github.com/user-attachments/assets/e7ce2abe-bbf5-449a-ae99-05f714ad3373)
> > ![Image](https://github.com/user-attachments/assets/319ec209-8b56-4805-8058-1835f9d6b314)
> 
> i have the same error, is there a solution?:(

The solution is very simple, just go to the official website to find the installation command and execute it.
```powershell
Invoke-WebRequest https://get.pnpm.io/install.ps1 -UseBasicParsing | Invoke-Expression
```

Although there is a solution, it's infuriating that the `pnpm self-update` command, which many people can't use, is still being shown as an upgrade prompt. The developers are being quite irresponsible, and that's what's most frustrating.

**Comment by Wycers** [comment]
same issue here

**Comment by YarosMallorca** [comment]
Same issue, and trying to update using the `Invoke-WebRequest https://get.pnpm.io/install.ps1 -UseBasicParsing | Invoke-Expression` command says I have the newest version

**Comment by m2paulc** [comment]
> [@PMA365](https://github.com/PMA365) How does this work? My pnpm.exe is in `/AppData/Local/pnpm/pnpm`. Installing pnpm via `npm` installs `pnpm` in `C:\Users\[username]\AppData\Roaming\npm`.
> 
> The only thing that worked for me was running the following in Powershell (admin), based on [@abcfy2](https://github.com/abcfy2) comment.
> 
> ```
> Invoke-WebRequest https://get.pnpm.io/install.ps1 -UseBasicParsing | Invoke-Expression
> ```

This has worked for me! pnpm updated to the latest version.

**Comment by Huodoo** [comment]
I simply downloaded the install script and modified it briefly, then renamed it pnpm-up and copied it to the $PNPM_HOME folder, then I could use pnpm-up instead of pnpm self-update

**Comment by you-hengh** [comment]
> I simply downloaded the install script and modified it briefly, then renamed it pnpm-up and copied it to the $PNPM_HOME folder, then I could use pnpm-up instead of pnpm self-update

I agree with you that I wrote a small command alias:
```powershell
function pnpm-up {
    Invoke-WebRequest https://get.pnpm.io/install.ps1 -UseBasicParsing | Invoke-Expression
}
```
Write it to your `$profile` file.

![Image](https://github.com/user-attachments/assets/6fef714f-9c05-4586-a4f4-d8a1553ab46b)

![Image](https://github.com/user-attachments/assets/853bd660-f1c2-4272-a03c-51f8e10203f2)

After saving the file, in a new terminal window, enter the command`pnpm-up`and your pnpm will be automatically updated

If this message has helped you, please like it so that more people can see it

**Comment by mutoe** [comment]
At first, I tried any of the above methods and none of them worked. However, when I tried to delete the `C:\Users\me\AppData\Local\pnpm` directory, I encountered the file being occupied.

My solution was:

1. unlock the directory using unlock file with locksmith, or find a way to kill all nodejs processes and pnpm processes
2. run `iwr https://get.pnpm.io/install.ps1 -useb | iex`

Now I'm up and running.

**Comment by kingma-sbw** [comment]
```
>pnpm self-update
Nothing to stop. No server is running for the store at C:\Users\Johannes\AppData\Local\pnpm\store\v10
 WARN  The target bin directory already contains an exe called pnpm, so removing C:\Users\Johannes\AppData\Local\pnpm\pnpm.EXE
 EPERM  EPERM: operation not permitted, unlink 'C:\Users\User\AppData\Local\pnpm\pnpm.EXE'
```


and 

```
> iwr https://get.pnpm.io/install.ps1 -useb | iex
Downloading pnpm from GitHub...

Running setup...

Copying pnpm CLI from C:\Users\User\AppData\Local\Temp\3a84ec1d-8dc6-4e9b-8b0d-4431fef56165\pnpm.exe to C:\Users\User\AppData\Local\pnpm\pnpm.exe
No changes to the environment were made. Everything is already up to date.
Remove-Item : Cannot remove item C:\Users\User\AppData\Local\Temp\3a84ec1d-8dc6-4e9b-8b0d-4431fef56165\pnpm.exe:
Access to the path 'C:\Users\User\AppData\Local\Temp\3a84ec1d-8dc6-4e9b-8b0d-4431fef56165\pnpm.exe' is denied.
At line:141 char:1
+ Remove-Item $tempFile
+ ~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (C:\Users\User\...f56165\pnpm.exe:FileInfo) [Remove-Item], Unauthoriz
   edAccessException
    + FullyQualifiedErrorId : RemoveFileSystemItemUnAuthorizedAccess,Microsoft.PowerShell.Commands.RemoveItemCommand
```

**Comment by uzairsyedahmed** [comment]
> ```
> >pnpm self-update
> Nothing to stop. No server is running for the store at C:\Users\User\AppData\Local\pnpm\store\v10
>  WARN  The target bin directory already contains an exe called pnpm, so removing C:\Users\User\AppData\Local\pnpm\pnpm.EXE
>  EPERM  EPERM: operation not permitted, unlink 'C:\Users\User\AppData\Local\pnpm\pnpm.EXE'
> ```
> 
> and
> 
> ```
> > iwr https://get.pnpm.io/install.ps1 -useb | iex
> Downloading pnpm from GitHub...
> 
> Running setup...
> 
> Copying pnpm CLI from C:\Users\User\AppData\Local\Temp\3a84ec1d-8dc6-4e9b-8b0d-4431fef56165\pnpm.exe to C:\Users\User\AppData\Local\pnpm\pnpm.exe
> No changes to the environment were made. Everything is already up to date.
> Remove-Item : Cannot remove item C:\Users\User\AppData\Local\Temp\3a84ec1d-8dc6-4e9b-8b0d-4431fef56165\pnpm.exe:
> Access to the path 'C:\Users\User\AppData\Local\Temp\3a84ec1d-8dc6-4e9b-8b0d-4431fef56165\pnpm.exe' is denied.
> At line:141 char:1
> + Remove-Item $tempFile
> + ~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : PermissionDenied: (C:\Users\User\...f56165\pnpm.exe:FileInfo) [Remove-Item], Unauthoriz
>    edAccessException
>     + FullyQualifiedErrorId : RemoveFileSystemItemUnAuthorizedAccess,Microsoft.PowerShell.Commands.RemoveItemCommand
> ```

I am having this issue too. I have tried everything that I can find on the internet. The windows powershell is running as an administrator so not sure what is happening.

**Comment by PMA365** [comment]
Just download the pnpm.exe file from the github and rename it to exactly pnpm.exe. then copy it manually to your last pnpm installation folder. for example for me is 
c:\Users\(myUserName)\AppDataLocal\pnpm
also add this address path to your system Environment path so you can use "pnpm" command every where in cmd or powershell.

**Comment by kudorgyozo** [comment]
what is the progress on this issue?

**Comment by amazfs** [comment]
> [@thakyZ](https://github.com/thakyZ) I don't have pnpm.cmd, just npm.cmd, is that what you meant?
> 
> Either way, I've switched to [using winget](https://pnpm.io/installation#using-winget) to manage pnpm versions and so far it's worked flawlessly.

only this worked for me!

**Comment by amazfs** [comment]
install a project version:  pnpm add - pnpm
uninstall the system version(if any):  winget uninstall pnpm
uninstall the global npm version(if any):  npm uninstall -g pnpm

**Comment by SkyLord2** [comment]
> here is the solution i tested:
> 
> 1. make sure node.js already installed
> 2. rename pnpm.exe to pnpm_.exe
> 3. pnpm_ -g i pnpm
> 4. make sure file pnpm.cmd appeared
> 5. delete pnpm_.exe
> 6. done. now pnpm manages itself

Wow，it‘s worked for me, thank you!

**Comment by wangrongding** [comment]
Wait for a best practice 👀

**Comment by square-story** [comment]
`iwr https://get.pnpm.io/install.ps1 -useb | iex`

use this to update

![Image](https://github.com/user-attachments/assets/5cd639a8-2fe0-41e7-8ec2-de3ba330d519)

**Comment by NT-AUTHORITY** [comment]
still not fixed

```
❯ pnpm self-update
Nothing to stop. No server is running for the store at C:\Users\NT\AppData\Local\pnpm\store\v10
 WARN  The target bin directory already contains an exe called pnpm, so removing C:\Users\NT\AppData\Local\pnpm\pnpm.EXE
 EPERM  EPERM: operation not permitted, unlink 'C:\Users\NT\AppData\Local\pnpm\pnpm.EXE'
```

![Image](https://github.com/user-attachments/assets/4f6e33c0-4716-42b7-a95c-7ed520fd717a)

i need to use the powershell method every time :(

```
iwr https://get.pnpm.io/install.ps1 -useb | iex
```

**Comment by codeflorist** [comment]
`ERR_PNPM_EPERM  EPERM: operation not permitted` is my personal nemesis by now.

**Comment by zkochan** [comment]
Maybe if we replace the `pnpm.exe` file in the PATH with a command shim that points to the actual pnpm.exe file, then `pnpm self-update` could replace the command shim with a new one that points to the new pnpm version. In this case the currently running pnpm executable will not be removed/replaced.
